File: heap/java/heap-sort/src/xyz/zrui/algorithm/heap/MaxHeap.java
package xyz.zrui.algorithm.heap;

public class MaxHeap<T extends Comparable> {

    protected T[] data;

    protected int count;

    protected int capacity;

    //构造函数， 构造一个空的堆，可容纳capacity个元素
    public MaxHeap(int capacity){

        this.data = (T[]) new Comparable[capacity+1];

        this.count = 0;

        this.capacity = capacity;
    }

    public int size(){
        return count;
    }

    public boolean isEmpty(){
        return count == 0;
    }

    public void insert(T t){
        data[count + 1] = t;
        count ++;
        shiftUp(count);
    }

    private void shiftUp(int k){

        while (k > 1 && data[k/2].compareTo(data[k]) < 0){
            swap(data,k,k/2);
            k /= 2;
        }

    }

    private void shiftDown(int k){

        while (2 * k <= count){

            int j = 2 * k;

            if (j + 1 <= count && data[j + 1].compareTo(data[j]) > 0){
                j += 1;
            }
            if (data[k].compareTo(data[j]) > 0){
                break;
            }
            swap(data,j,k);
            k = j;
        }



    }

    public T extractMax(){
        T ret = data[1];
        swap(data,count,1);
        count --;
        shiftDown(1);
        return ret;
    }

    /**
     * 交换数组间的位置
     *
     * @param arr 目标数组
     * @param i   交换位置索引
     * @param j   交换位置索引
     */
    private  void swap(Object[] arr, int i, int j) {
        Object t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }


    // 测试 MaxHeap
    public static void main(String[] args) {

        MaxHeap<Integer> maxHeap = new MaxHeap<Integer>(100);
        int N = 50; // 堆中元素个数
        int M = 100; // 堆中元素取值范围[0, M)
        for( int i = 0 ; i < N ; i ++ )
            maxHeap.insert( new Integer((int)(Math.random() * M)) );
        System.out.println(maxHeap.size());

    }


}

File: heap/python/max_heap.py
# -*- coding: utf-8 -*

'''
                        最大堆实现
                           98
                /                      \
               96                      84
          /          \            /          \
         92          82          78          47
       /    \      /    \      /    \      /    \
      33    26    51    85    50    15    44    60
     / \   / \   / \   / \   / \   / \   / \   / \
    40 51 98 51 7  17 94 82 32 21 64 60 7  44 63 63
'''
import random

class Maxheap(object):

    def __init__(self,cpacity,arr = None):

        self.data = [None] * (cpacity + 1)
        self.cpacity = cpacity
        if arr is None:
            self.count = 0
        else:
            for i in range(0,cpacity):
                self.data[i + 1]= arr[i]
            self.count = cpacity;
            for i in range(self.count / 2, 0, -1):
                self.__shifDown(i)

    def size(self):
        return self.count

    def isEmpty(self):
        return self.count == 0

    def __shiftUp(self,k):

        while k > 1 and self.data[k] > self.data[k / 2]:
            self.data[k],self.data[k / 2] = self.data[k / 2], self.data[k]
            k /= 2


    def insert(self,data):
        self.data[self.count + 1] = data
        self.count += 1
        self.__shiftUp(self.count)

    def __shifDown(self,k):
        while k * 2 <= self.count:

            j = k * 2

            if self.count >= j + 1 and self.data[j + 1] > self.data[j]:
                j += 1

            if self.data[k] > self.data[j]:
                break

            self.data[k], self.data[j] = self.data[j],self.data[k]

            k = j

    def extractMax(self):
        ret = self.data[1]
        self.data[1], self.data[self.count] = self.data[self.count], self.data[1]
        self.count -= 1
        self.__shifDown(1)
        return ret






if __name__ == '__main__':

    N = 31
    M = 100
    heap = Maxheap(N)
    for i in range(0,N):
        k = random.randint(1, M)
        heap.insert(k)

    # arr = [random.randint(1,M) for i in range(N)]
    # heap = Maxheap(len(arr),arr)
#[None, 95, 84, 85, 78, 80, 50, 74, 62, 70, 59, 51, 38, 47, 57, 71, 9, 16, 19, 53, 18, 25, 44, 7, 28, 2, 24, 16, 18, 13, 33, 70]
    print(heap.size())
    print(heap.data)
    print(heap.extractMax())
